src/arch/riscv/irqc/aia/inc/aplic.h

```diff
@@ -0,0 +1,269 @@
+#define APLIC_DOMAINCFG_DM              (1U << 2)
+#define APLIC_DOMAINCFG_IE              (1U << 8)
+#define APLIC_DOMAINCFG_RO80            (0x80 << 24)
+
+#define APLIC_SRCCFG_D                  (1U << 10)
+#define APLIC_SRCCFG_SM                 ((1U << 0) | (1U << 1) | (1U << 2))
+
+#define APLIC_TARGET_HART_IDX_SHIFT     (18)
+#define APLIC_TARGET_HART_IDX_MASK      (0x3FFF)
+#define APLIC_TARGET_IPRIO_MASK         (0xFF)
+#define APLIC_TARGET_PRIO_DEFAULT       (1)
+#define APLIC_TARGET_MASK               (0xFFFC00FF)
+/** Data structures for APLIC devices */
+struct irqc_global_hw {
+    uint32_t domaincfg;
+    uint32_t sourcecfg[APLIC_NUM_SRCCFG_REGS];
+    uint8_t  reserved1[0x1C00 - 0x1000];
+    uint32_t setip[APLIC_NUM_SETIx_REGS];
+    uint8_t  reserved2[0x1CDC - 0x1C80];
+    uint32_t setipnum;
+    uint8_t  reserved3[0x1D00 - 0x1CE0];
+    uint32_t in_clrip[APLIC_NUM_CLRIx_REGS];
+    uint8_t  reserved4[0x1DDC - 0x1D80];
+    uint32_t clripnum;
+    uint8_t  reserved5[0x1E00 - 0x1DE0];
+    uint32_t setie[APLIC_NUM_SETIx_REGS];
+    uint8_t  reserved6[0x1EDC - 0x1E80];
+    uint32_t setienum;
+    uint8_t  reserved7[0x1F00 - 0x1EE0];
+    uint32_t clrie[APLIC_NUM_CLRIx_REGS];
+    uint8_t  reserved8[0x1FDC - 0x1F80];
+    uint32_t clrienum;
+    uint8_t  reserved9[0x2000 - 0x1FE0];
+    uint32_t setipnum_le;
+    uint32_t setipnum_be;
+    uint8_t reserved10[0x3000 - 0x2008];
+    uint32_t genmsi;
+    uint32_t target[APLIC_NUM_TARGET_REGS];
+} __attribute__((__packed__, aligned(PAGE_SIZE)));
+
+struct irqc_hart_hw {
+    uint32_t idelivery;
+    uint32_t iforce;
+    uint32_t ithreshold;
+    uint8_t  reserved[0x18-0x0C];
+    uint32_t topi;
+    uint32_t claimi;
+}__attribute__((__packed__, aligned(APLIC_IDC_SIZE))); // IDC structure CANNOT be page aligned.
+
+extern uint32_t impl_src[APLIC_MAX_INTERRUPTS];
+
+extern volatile struct irqc_global_hw *irqc_global;
+extern volatile struct irqc_hart_hw *irqc_hart;
+
+/** Initialization Functions */
+/**
+ * @brief Initialize the aplic domain.
+ * 
+ */
+void aplic_init(void);
+
+/**
+ * @brief Initialize the aplic IDCs. 
+ * The IDC component is the closest to the cpu.
+ * 
+ */
+void aplic_idc_init(void);
+
+/** Domain functions */
+/**
+ * @brief Write to aplic's sourcecfg register
+ * 
+ * @param int_id interruption ID identifies the interrupt to be configured/read.
+ * @param val Value to be written into sourcecfg
+ */
+void aplic_set_sourcecfg(irqid_t int_id, uint32_t val);
+
+/**
+ * @brief Read from aplic's sourcecfg register
+ * 
+ * @param int_id interruption ID identifies the interrupt to be configured/read.
+ * @return uint32_t 32 bit value containing interrupt int_id's configuration.
+ */
+uint32_t aplic_get_sourcecfg(irqid_t int_id);
+
+/**
+ * @brief Set a given interrupt as pending, using setipnum register.
+ * This should be faster than aplic_set_pend.
+ * 
+ * @param int_id Interrupt to be set as pending
+ */
+void aplic_set_pend(irqid_t int_id);
```

Comment:
Similar to what we have in plic, we could probably merge the `aplic_set_pend` and `aplic_clr_pend` functions by adding `bool val` parameter.

What do you think? Should I do it, or as it is it's fine?